Fix/api

ldg-96 opened the pull request:
api를 사용할때 재사용성, 확장성을 고려해서 설계를 변경했습니다.

ldg-96 committed:
API Layer 재설계 (#75)

ldg-96 committed:
API Layer 재설계 (#75)

ldg-96 commented:
+추가질문 있습니다.
cloudfront 캐쉬무효화 관련 질문입니다. 제가 이해한 바로는 월1000건 무료인대 cloudfront는 경로 수 단위로 비용을 계산하는것 같습니다.
예전에 리뷰해주실때 캐쉬무효화시 /* 전체경로에 대해 하지말고 변경된 파일들의 경로에 대해서만 무효화를 진행하라고 하셨던 기억이 있는데 어떤 이유가 있을까요?

SimYunSup commented:
@ldg-96 https://aws.amazon.com/ko/blogs/networking-and-content-delivery/host-single-page-applications-spa-with-tiered-ttls-on-cloudfront-and-s3/

ldg-96 commented:
윤섭님이 보내주신 링크글에 대하여 읽어보았습니다. 글의 내용은 계층형 TTL을 사용하여 캐시를 건드리지 않고 새롭게 배포하는 방식인것 같습니다.
- 정적 번들, 이미지는 빌드 과정에서 해시가 들어간 이름으로 생성하고 S3업로드시 캐시 기간을 길게잡습니다.
- index.html같은 진입점은 매 요청마다 최신 여부를 검증할 수 있도록 Cache-Control: max-age=0  (문서에는 60초 였지만 저는 0초로 두어 바로바로 최신 여부를 검증해보도록 했습니다.) 으로 둡니다.

`env:
  LONG_CACHE: "--cache-control public,max-age=31536000,immutable"
  SHORT_CACHE: "--cache-control public,max-age=0,no-cache,must-revalidate"
  S3_SYNC_OPTS: "--delete --checksum"
`

yml에서 위와같이 변수를 두고
![image](https://github.com/user-attachments/assets/a24b7acf-c4ec-4c3d-b525-4f020d30a4f2)
이미지와 같이 적용하여보았습니다.

SimYunSup commented on src/utils/apiHandler.ts:
```diff
@@ -0,0 +1,13 @@
+export async function apiHandler<T>(fn: () => Promise<T>): Promise<T> {
+  try {
+    return await fn()
+  } catch (err) {
+    let message = '백엔드 통신 실패'
+    if (err instanceof Error) {
+      message = err.message
+    } else if (typeof err === 'string') {
+      message = err
+    }
+    throw new Error(message)
+  }
+}
```
1. `throw string`하는 경우도 별로 없을 뿐더러, 만약 에러 포맷을 일관화 + 타입관리하고 싶으셨다면 `try...catch`로 하시는 것은 `err` 타입 가드가 되지 않아 더욱 타입 관리가 안되어 보이는데, 굳이 `try...catch` 구문을 두 번(이 함수에서 한번, 밖에서 한번.) 써야 하는지 모르겠습니다.
2. DRY 원칙은 말 그대로 원칙입니다. 원칙을 엄청나게 고수하는 것 또한 비용이 매우 많이 듧니다.

ldg-96 committed:
API Layer 재설계 (#75)
수정내용 : ApiError 타입을 지정해서 err타입을 관리하도록 하였으며 apiHandler를 삭제하고 try...catch구조를 한번만 사용하는식으로 변경하였습니다

ldg-96 committed:
Merge branch 'develop' into fix/api
